nodes.py

```diff
@@ -32,27 +32,50 @@ def gguf_sd_loader_get_orig_shape(reader, tensor_name):
         raise TypeError(f"Bad original shape metadata for {field_key}: Expected ARRAY of INT32, got {field.types}")
     return torch.Size(tuple(int(field.parts[part_idx][0]) for part_idx in field.data))
 
-def gguf_sd_loader(path):
+def gguf_sd_loader(path, handle_prefix="model.diffusion_model.", require_prefix=False):
     """
     Read state dict as fake tensors
     """
     reader = gguf.GGUFReader(path)
     sd = {}
     dt = {}
+    arch_field = reader.get_field("general.architecture")
+    if arch_field is not None:
+        if len(arch_field.types) != 1 or arch_field.types[0] != gguf.GGUFValueType.STRING:
+            raise TypeError(f"Bad type for GGUF general.architecture key: expected string, got {arch_field.types!r}")
+        arch_str = str(arch_field.parts[arch_field.data[-1]], encoding="utf-8")
+        if arch_str not in {"flux", "sd1", "sdxl"}:
+            raise ValueError(f"Unexpected architecture type in GGUF file, expected one of flux, sd1, sdxl but got {arch_str!r}")
+    else:
+        arch_str = None
+    if handle_prefix is not None:
+        prefix_len = len(handle_prefix)
+        tensor_names = set(tensor.name for tensor in reader.tensors)
+        has_prefix = any(s.startswith(handle_prefix) for s in tensor_names)
+    else:
+        has_prefix = False
+    if require_prefix and not has_prefix:
+        raise RuntimeError(f"Loader requires prefix {handle_prefix!r} which does not exist in the model")

```

Comment:
I don't think we need "require_prefix" as I'm vehemently against ever supporting multiple models in a single GGUF file as that's 100% against what the format specs intend.

I think we can just use this same bit of code to strip out the unet with the prefix hardcoded as "model.diffusion_model." since [even SD3 uses that same prefix](https://huggingface.co/stabilityai/stable-diffusion-3-medium/tree/main?show_file_info=sd3_medium.safetensors) - we can turn it into a loop and add more if that ever changes.

https://github.com/city96/ComfyUI-GGUF/blob/3d9869240bd4ff786030ad85b09441045bb3fda9/tools/convert.py#L62-L70

Reply:
> I don't think we need "require_prefix" as I'm vehemently against ever supporting multiple models in a single GGUF file as that's 100% against what the format specs intend.

i can remove it if you insist but i _really_ hope you'll reconsider. combined checkpoints are way, way easier to deal with from a user perspective. splitting your existing checkpoints into separate VAE, CLIP and u-net and then having to quantize them individually is both a hassle to do and harder to manage even after they've been quantized.

also the `stable-diffusion.cpp` GGUF file already has VAE, CLIP and u-net combined: you can't make them stop doing that. the most you could do by removing support here is make it some the user is only able to use the u-net part and the rest is just useless to them. they'll have to do unnecessary work to get VAE/CLIP that already exists in the GGUF file.

Reply:
Pretty sure almost all models use the same 4 VAE models and the same 3-4 CLIP encoders anyway. Is there really any point in copying these? Besides, we'd then have to deal with the VAE model (which is like 300MB and probably shouldn't be quantized anyway, since most of it's memory usage is during runtime).

Then you'd have the issue about people being confused if a given gguf file is the actual unet or a combined checkpoint one. "which loader do I use" etc. With llama.cpp, even the mmproj model is split as a separate file for vllms.

Besides, this repo remains to mostly be centered around large models like flux and the upcoming sd3.1 (assuming it'll be larger than sd3). With those, packaging the T5 model would just result in comically large checkpoints. And I already know people would start packaging them and releasing crap like "flux1-dev-schnell-merge_q4_k_s_with_T5_q5_k_m.gguf" given the chance lol.

![image](https://github.com/user-attachments/assets/f9d5e41a-2bf6-4536-9967-c70c73250bbb)

Reply:
okay, i removed `require_prefix`. stuff like custom VAEs (and trained TE) is more common for SD15 models. i think it would be good from a user perspective if they could treat `.gguf` files similarly to `.safetensors` but it's your project.

leaving the prefix code relatively generic rather than hardcoding it doesn't increase the complexity much: i left that as is. if you ever do want to support different prefixes later on it will be easier to add, otherwise you can just never pass that parameter and it be effectively the same as hardcoding `model.diffusion_model.`

Reply:
Yeah sure that's fine, we'll probably have to move that loader function anyway assuming I ever get the time to work on the c++ stuff.

Reply:
there are GGML bindings for Python. if it was me, what i'd probably do is look to make my own conversion/quantization script and completely avoid having to mess with llama.cpp, patch GGML tensor name length, etc. that would be somewhat more up front effort but maintaining patches is basically unbounded effort.